typings/styled-components.d.ts

```diff
@@ -24,10 +29,29 @@ export interface InterpolationFunction<P> {
   (props: P): Interpolation<P>;
 }
 
-export interface ThemedStyledFunction<P, T> {
-  (strings: TemplateStringsArray, ...interpolations: Interpolation<ThemedStyledProps<P, T>>[]): ComponentClass<ThemedOuterStyledProps<P, T>>;
-  <U>(strings: TemplateStringsArray, ...interpolations: Interpolation<ThemedStyledProps<P & U, T>>[]): ComponentClass<ThemedOuterStyledProps<P & U, T>>;
+type WithComponentOverloads<Tags, T> = {
+  [K in keyof Tags]: StyledComponentClass<Tags[K], T>;
+};
+
+type Attrs<P, A extends Partial<P>, T> = {
+  [K in keyof A]: ((props: ThemedStyledProps<P, T>) => A[K]) | A[K];
+};
+
+export interface StyledComponentClass<P, T, O = P> extends ComponentClass<ThemedOuterStyledProps<P, T>> {
+  extend: ThemedStyledFunction<P, T>;
+
+  withComponent<K extends keyof HTMLTags>(tag: K): WithComponentOverloads<HTMLTags, T>[K];
+  withComponent<K extends keyof SVGTags>(tag: K): WithComponentOverloads<SVGTags, T>[K];
+  withComponent(element: ComponentClass<P>): StyledComponentClass<P, T>;
+  attrs<U, A  extends Partial<P> = {}>(attrs: Attrs<P | U, A, T>): ThemedStyledFunction<P & A & U, T, O & U>;
```

Comment:
No, it has to be exactly that `Attrs<P & U, A, T>`
`U` represents a newly introduced fresh props. They will be added to props accessible in the attr's functions.